README.md
# tap-github

This is a [Singer](https://singer.io) tap that produces JSON-formatted
data from the GitHub API following the [Singer
spec](https://github.com/singer-io/getting-started/blob/master/docs/SPEC.md).

This tap:
- Pulls raw data from the [GitHub REST API](https://developer.github.com/v3/)
- Extracts the following resources from GitHub for a single repository:
  - [Assignees](https://docs.github.com/en/rest/reference/issues#list-assigneess)
  - [Collaborators](https://docs.github.com/en/rest/reference/repos#list-repository-collaborators)
  - [Commits](https://docs.github.com/en/rest/reference/repos#list-commits)
  - [Commit Comments](https://docs.github.com/en/rest/reference/repos#list-commit-comments-for-a-repository)
  - [Events](https://docs.github.com/en/rest/reference/issues#events)
  - [Issues](https://docs.github.com/en/rest/reference/issues#list-repository-issues)
  - [Issue Events](https://docs.github.com/en/rest/reference/issues#list-issue-events-for-a-repository)
  - [Issue Milestones](https://docs.github.com/en/rest/reference/issues#list-milestones)
  - [Projects](https://docs.github.com/en/rest/reference/projects#list-repository-projects)
  - [Project Cards](https://docs.github.com/en/rest/reference/projects#list-project-cards)
  - [Project Columns](https://docs.github.com/en/rest/reference/projects#list-project-columns)
  - [Pull Requests](https://docs.github.com/en/rest/reference/pulls#list-pull-requests)
  - [PR Commits](https://docs.github.com/en/rest/reference/pulls#list-commits-on-a-pull-request)
  - [Releases](https://docs.github.com/en/rest/reference/repos#list-releases)
  - [Comments](https://docs.github.com/en/rest/reference/issues#list-issue-comments-for-a-repository)
  - [Reviews](https://docs.github.com/en/rest/reference/pulls#list-reviews-for-a-pull-request)
  - [Review Comments](https://docs.github.com/en/rest/reference/pulls#list-review-comments-in-a-repository)
  - [Stargazers](https://docs.github.com/en/rest/reference/activity#list-stargazers)
  - [Teams](https://docs.github.com/en/rest/reference/teams#list-teams)
  - [Team Members](https://docs.github.com/en/rest/reference/teams#list-team-members)
  - [Team Memberships](https://docs.github.com/en/rest/reference/teams#get-team-membership-for-a-user)
- Outputs the schema for each resource
- Incrementally pulls data based on the input state

## Quick start

1. Install

   We recommend using a virtualenv:

    ```bash
    > virtualenv -p python3 venv
    > source venv/bin/activate
    > pip install tap-github
    ```

2. Create a GitHub access token

    Login to your GitHub account, go to the
    [Personal Access Tokens](https://github.com/settings/tokens) settings
    page, and generate a new token with at least the `repo` scope. Save this
    access token, you'll need it for the next step.

3. Create the config file

    Create a JSON file containing the start date, access token you just created
    and the path to one or multiple repositories that you want to extract data from. Each repo path should be space delimited. The repo path is relative to `"base_url"`
    (Default: `https://github.com/`). For example the path for this repository is
    `singer-io/tap-github`. You can also add request timeout to set the timeout for requests which is an optional parameter with default value of 300 seconds.

    ```json
    {
      "access_token": "your-access-token",
      "repository": "singer-io/tap-github singer-io/getting-started",
      "start_date": "2021-01-01T00:00:00Z",
      "request_timeout": 300,
      "base_url": "https://api.github.com"
    }
    ```
4. Run the tap in discovery mode to get properties.json file

    ```bash
    tap-github --config config.json --discover > properties.json
    ```
5. In the properties.json file, select the streams to sync

    Each stream in the properties.json file has a "schema" entry.  To select a stream to sync, add `"selected": true` to that stream's "schema" entry.  For example, to sync the pull_requests stream:
    ```
    ...
    "tap_stream_id": "pull_requests",
    "schema": {
      "selected": true,
      "properties": {
        "updated_at": {
          "format": "date-time",
          "type": [
            "null",
            "string"
          ]
        }
    ...
    ```

6. Run the application

    `tap-github` can be run with:

    ```bash
    tap-github --config config.json --properties properties.json
    ```

---

Copyright &copy; 2018 Stitch


CHANGELOG.md
# Changelog

# 2.0.0
  * Schema updates [#170](https://github.com/singer-io/tap-github/pull/170) [#169](https://github.com/singer-io/tap-github/pull/169)
    * Update data types of fields in `events` and `issue_events` stream
    * Add missing fields to the schemas
  * Update dict based implementation to class based [#168](https://github.com/singer-io/tap-github/pull/168)
  * Implement currently syncing for repos and streams [#171](https://github.com/singer-io/tap-github/pull/171) [#174](https://github.com/singer-io/tap-github/pull/174)
  * Implement custom exception handling and backoff for 5xx error [#166](https://github.com/singer-io/tap-github/pull/166)
  * Support of custom domain [#172](https://github.com/singer-io/tap-github/pull/172)
  * Sync teams at organization level [#173](https://github.com/singer-io/tap-github/pull/173) 
  * Update integration test suite [#167](https://github.com/singer-io/tap-github/pull/167)

# 1.10.4
  * Fix team_members stream primary Key [#157] (https://github.com/singer-io/tap-github/pull/157)

# 1.10.3
  * Implemented wildcard implementation [#145] (https://github.com/singer-io/tap-github/pull/145)
  * Added additional test coverage [#145] (https://github.com/singer-io/tap-github/pull/145)

# 1.10.2
  * Added Request Timeout

# 1.10.1
  * Add support of object in "parent" field for teams [149](https://github.com/singer-io/tap-github/pull/149)

# 1.10.0
  * Handle rate limiting [#113](https://github.com/singer-io/tap-github/pull/113)
  * Handle `None` date times in the Issue Milestones Stream [#114](https://github.com/singer-io/tap-github/pull/114)
  * Change the Stargazers Stream to be Full Table replication [#118](https://github.com/singer-io/tap-github/pull/118)
  * Log an error message if an organization cannot be found [#121](https://github.com/singer-io/tap-github/pull/121)
  * Bump `singer-python` dependency to `v5.12.1` [#117](https://github.com/singer-io/tap-github/pull/117)
  * Remove stream `pull_request_reviews` [#117](https://github.com/singer-io/tap-github/pull/117)
  * Add API access check to discovery mode [#123](https://github.com/singer-io/tap-github/pull/123)

# 1.9.2
  * Adds `base` to `pull_requests` schema [#109](https://github.com/singer-io/tap-github/pull/109)

# 1.9.1
  * Fix some field data types for the `issue_events` stream [#102](https://github.com/singer-io/tap-github/pull/102)
    * `issue_events.issue.milestone`
    * `issue_events.issue.active_lock_reason`
    * `issue_events.issue.performed_via_github_app`
    * `issue_events.issue.labels.description`
    * `issue_events.performed_via_github_app`

# 1.9.0
  * Adds `issue_events` stream [#92](https://github.com/singer-io/tap-github/pull/92)
  * Makes `project_cards` stream a child of `project_columns`[#89](https://github.com/singer-io/tap-github/pull/89)

# 1.8.2
  * Fix [issue 83](https://github.com/singer-io/tap-github/issues/83) by removing `format: date-time` from `html_url` field in `projects.json`.

# 1.8.1
  * Add stream `pr_commits` back in [#81](https://github.com/singer-io/tap-github/pull/81)

# 1.8.0
  * Added team_memberships stream [#75](https://github.com/singer-io/tap-github/pull/75)

# 1.7.1
  * Remove `"format": "date-time"` from JSON schema's for keys that are not dates [#72](https://github.com/singer-io/tap-github/pull/72)

# 1.7.0
  * Adds many new streams [#70](https://github.com/singer-io/tap-github/pull/70)

# 1.6.0
  * Adds a new stream for PR Commits [#67](https://github.com/singer-io/tap-github/pull/67)

# 1.4.0
  * Adds a new stream for Github Reviews [#53](https://github.com/singer-io/tap-github/pull/53)

# 1.3.6
  * Add a call to `raise_for_status()` to catch more errors than before [#55](https://github.com/singer-io/tap-github/pull/55)
  * Update schemas to make some fields `date-time`s [#60](https://github.com/singer-io/tap-github/pull/60)

## 1.3.5
  * Add the `submitted_at` field to the `review` stream [#52](https://github.com/singer-io/tap-github/pull/52)

## 1.3.4
  * Filter pull_requests stream using its updated at [#49](https://github.com/singer-io/tap-github/pull/49)
  * Fix a bug in the setup of sub streams [#47](https://github.com/singer-io/tap-github/pull/47)

## 1.3.3
  * Include closed issues when we retrieve issues [#45](https://github.com/singer-io/tap-github/pull/45)

## 1.3.2
  * Checks responses for an auth error (403) and throws exception [#39](https://github.com/singer-io/tap-github/pull/39)
  * Remove `milestone` and add type to `closed_at` in `Issues` schema [#40](https://github.com/singer-io/tap-github/pull/40)

## 1.3.1
  * Adds empty schema for issue.assignee [#37](https://github.com/singer-io/tap-github/pull/37)

## 1.3.0
  * Add functionality to accept multiple repos [#33](https://github.com/singer-io/tap-github/pull/33)

## 1.2.1
  * Update version of `requests` to `2.20.0` in response to CVE 2018-18074

## 1.2.0
  * Adds issue comments stream  [#32](https://github.com/singer-io/tap-github/pull/32)

## 1.1.1
  * Rename the comments stream to review_comments [#31](https://github.com/singer-io/tap-github/pull/31)

## 1.1.0
  * Added the comments stream [#30](https://github.com/singer-io/tap-github/pull/30)

## 1.0.5
  * More updates to the issues schema

## 1.0.4
  * Updates the issues schema [#29](https://github.com/singer-io/tap-github/pull/29)

## 1.0.3
  * Adds additional schema to the issues JSON schema [#28](https://github.com/singer-io/tap-github/pull/28)

## 1.0.2
  * Checks responses for an not found error (404) and throws exception

## 1.0.1
  * Checks responses for an auth error (401) and throws exception [#26](https://github.com/singer-io/tap-github/pull/26)

## 1.0.0
  * Writes appropriate metadata to allow for selection of fields
  * Adds bookmarking and incremental replication to Commits and Issues streams [#22](https://github.com/singer-io/tap-github/pull/22)

## 0.5.4

tests/test_github_sync.py
from tap_tester import connections
from base import TestGithubBase

class TestGithubSync(TestGithubBase):
    """Test tap sync mode and metadata conforms to standards."""

    @staticmethod
    def name():
        return "tap_tester_github_sync_test"

    def get_properties(self):

        return {
            'start_date' : '2021-10-01T00:00:00Z',
            'base_url': 'https://api.github.com',
            'repository': 'singer-io/test-repo'
        }

    def test_run(self):
        """
        Testing that sync creates the appropriate catalog with valid metadata.
        Verify that all fields and all streams have selected set to True in the metadata
        """
        conn_id = connections.ensure_connection(self)

        found_catalogs = self.run_and_verify_check_mode(conn_id)

        self.perform_and_verify_table_and_field_selection(conn_id,found_catalogs)

        record_count_by_stream = self.run_and_verify_sync(conn_id)

        self.assertGreater(sum(record_count_by_stream.values()), 0)


tests/unittests/test_main.py
import unittest
from unittest import mock
from tap_github import main
from tap_github.discover import discover

class MockArgs:
    """Mock args object class"""
    
    def __init__(self, config = None, properties = None, state = None, discover = False) -> None:
        self.config = config 
        self.properties = properties
        self.state = state
        self.discover = discover

@mock.patch("tap_github.GithubClient")
@mock.patch("singer.utils.parse_args")
class TestDiscoverMode(unittest.TestCase):
    """
    Test main function for discover mode
    """

    mock_config = {"start_date": "", "access_token": ""}
    
    @mock.patch("tap_github._discover")
    def test_discover_with_config(self, mock_discover, mock_args, mock_verify_access):
        """Test `_discover` function is called for discover mode"""
        mock_discover.return_value = dict()
        mock_args.return_value = MockArgs(discover = True, config = self.mock_config)
        main()
        
        self.assertTrue(mock_discover.called)


@mock.patch("tap_github.GithubClient")
@mock.patch("singer.utils.parse_args")
@mock.patch("tap_github._sync")
class TestSyncMode(unittest.TestCase):
    """
    Test main function for sync mode
    """

    mock_config = {"start_date": "", "access_token": ""}
    mock_catalog = {"streams": [{"stream": "teams", "schema": {}, "metadata": {}}]}

    @mock.patch("tap_github._discover")
    def test_sync_with_properties(self, mock_discover, mock_sync, mock_args, mock_client):
        """Test sync mode with properties given in args"""

        mock_client.return_value = "mock_client"
        mock_args.return_value = MockArgs(config=self.mock_config, properties=self.mock_catalog)
        main()
        
        # Verify `_sync` is called with expected arguments
        mock_sync.assert_called_with("mock_client", self.mock_config, {}, self.mock_catalog)
        
        # verify `_discover` function is not called
        self.assertFalse(mock_discover.called)

    @mock.patch("tap_github._discover")
    def test_sync_without_properties(self, mock_discover, mock_sync, mock_args, mock_client):
        """Test sync mode without properties given in args"""
        
        mock_discover.return_value = {"schema": "", "metadata": ""}
        mock_client.return_value = "mock_client"
        mock_args.return_value = MockArgs(config=self.mock_config)
        main()
        
        # Verify `_sync` is called with expected arguments
        mock_sync.assert_called_with("mock_client", self.mock_config, {}, {"schema": "", "metadata": ""})

        # verify `_discover` function is  called
        self.assertTrue(mock_discover.called)

    def test_sync_with_state(self, mock_sync, mock_args, mock_client):
        """Test sync mode with state given in args"""
        mock_state = {"bookmarks": {"projec ts": ""}}
        mock_client.return_value = "mock_client"
        mock_args.return_value = MockArgs(config=self.mock_config, properties=self.mock_catalog, state=mock_state)
        main()
        
        # Verify `_sync` is called with expected arguments
        mock_sync.assert_called_with("mock_client", self.mock_config, mock_state, self.mock_catalog)

@mock.patch("tap_github.GithubClient")
class TestDiscover(unittest.TestCase):
    """Test `discover` function."""
    
    def test_discover(self, mock_client):
        
        return_catalog = discover(mock_client)
        
        self.assertIsInstance(return_catalog, dict)

    @mock.patch("tap_github.discover.Schema")
    @mock.patch("tap_github.discover.LOGGER.error")
    def test_discover_error_handling(self, mock_logger, mock_schema, mock_client):
        """Test discover function if exception arises."""
        mock_schema.from_dict.side_effect = [Exception]
        with self.assertRaises(Exception):
            discover(mock_client)

        # Verify logger called 3 times when an exception arises.
        self.assertEqual(mock_logger.call_count, 3)


tap_github/schema.py
import os
import json
from singer import metadata
import singer
from tap_github.streams import STREAMS

def get_abs_path(path):
    """
    Get the absolute path for the schema files.
    """
    return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)

def load_schema_references():
    """
    Load the schema files from the schema folder and return the schema references.
    """
    shared_schema_path = get_abs_path('schemas/shared')

    shared_file_names = [f for f in os.listdir(shared_schema_path)
                         if os.path.isfile(os.path.join(shared_schema_path, f))]

    refs = {}
    for shared_schema_file in shared_file_names:
        with open(os.path.join(shared_schema_path, shared_schema_file)) as data_file:
            refs['shared/' + shared_schema_file] = json.load(data_file)

    return refs

def get_schemas():
    """
    Load the schema references, prepare metadata for each streams and return schema and metadata for the catalog.
    """
    schemas = {}
    field_metadata = {}

    refs = load_schema_references()
    for stream_name, stream_metadata in STREAMS.items():
        schema_path = get_abs_path('schemas/{}.json'.format(stream_name))

        with open(schema_path) as file:
            schema = json.load(file)

        schemas[stream_name] = schema
        schema = singer.resolve_schema_references(schema, refs)

        mdata = metadata.new()
        mdata = metadata.get_standard_metadata(
                schema=schema,
                key_properties = (hasattr(stream_metadata, 'key_properties') or None) and stream_metadata.key_properties,
                valid_replication_keys = (hasattr(stream_metadata, 'replication_keys') or None) and stream_metadata.replication_keys,
                replication_method = (hasattr(stream_metadata, 'replication_method') or None) and stream_metadata.replication_method
            )
        mdata = metadata.to_map(mdata)

        # Loop through all keys and make replication keys and primary keys of child stream which are not automatic in parent stream of automatic inclusion
        for field_name in schema['properties'].keys():

            pk_child_fields = (hasattr(stream_metadata, 'pk_child_fields') or None) and stream_metadata.pk_child_fields
            replication_keys = (hasattr(stream_metadata, 'replication_keys') or None) and stream_metadata.replication_keys
            if (replication_keys and field_name in replication_keys) or (pk_child_fields and field_name in pk_child_fields):
                mdata = metadata.write(mdata, ('properties', field_name), 'inclusion', 'automatic')


        mdata = metadata.to_list(mdata)
        field_metadata[stream_name] = mdata


    return schemas, field_metadata


tap_github/__init__.py
import json
import sys
import singer
from tap_github.discover import discover as _discover
from tap_github.client import GithubClient
from tap_github.sync import sync as _sync

LOGGER = singer.get_logger()

REQUIRED_CONFIG_KEYS = ['start_date', 'access_token', 'repository']

def do_discover(client):
    """
    Call the discovery function.
    """
    catalog = _discover(client)
    # Dump catalog
    json.dump(catalog, sys.stdout, indent=2)

@singer.utils.handle_top_exception(LOGGER)
def main():
    """
    Run discover mode or sync mode.
    """
    args = singer.utils.parse_args(REQUIRED_CONFIG_KEYS)

    config = args.config

    client = GithubClient(config)

    state = {}
    if args.state:
        state = args.state

    if args.discover:
        do_discover(client)
    else:
        catalog = args.properties if args.properties else _discover(client)
        _sync(client, config, state, catalog)

if __name__ == '__main__':
    main()


tap_github/discover.py
import singer
from singer.catalog import Catalog, CatalogEntry, Schema
from tap_github.schema import get_schemas

LOGGER = singer.get_logger()

def discover(client):
    """
    Run the discovery mode, prepare the catalog file and return catalog.
    """
    # Check credential in the discover mode.
    client.verify_access_for_repo()

    schemas, field_metadata = get_schemas()
    catalog = Catalog([])

    for stream_name, schema_dict in schemas.items():
        try:
            schema = Schema.from_dict(schema_dict)
            mdata = field_metadata[stream_name]
        except Exception as err:
            LOGGER.error(err)
            LOGGER.error('stream_name: %s', stream_name)
            LOGGER.error('type schema_dict: %s', type(schema_dict))
            raise err

        key_properties = mdata[0]['metadata'].get('table-key-properties')
        catalog.streams.append(CatalogEntry(
            stream=stream_name,
            tap_stream_id=stream_name,
            key_properties= key_properties,
            schema=schema,
            metadata=mdata
        ))

    return catalog.to_dict()


tap_github/client.py
import time
import requests
import backoff
from simplejson import JSONDecodeError
import singer
from singer import metrics

LOGGER = singer.get_logger()
DEFAULT_SLEEP_SECONDS = 600
DEFAULT_DOMAIN = "https://api.github.com"

# Set default timeout of 300 seconds
REQUEST_TIMEOUT = 300

class GithubException(Exception):
    pass

class Server5xxError(GithubException):
    pass

class BadCredentialsException(GithubException):
    pass

class AuthException(GithubException):
    pass

class NotFoundException(GithubException):
    pass

class BadRequestException(GithubException):
    pass

class InternalServerError(Server5xxError):
    pass

class UnprocessableError(GithubException):
    pass

class NotModifiedError(GithubException):
    pass

class MovedPermanentlyError(GithubException):
    pass

class ConflictError(GithubException):
    pass

class RateLimitExceeded(GithubException):
    pass

class TooManyRequests(GithubException):
    pass


ERROR_CODE_EXCEPTION_MAPPING = {
    301: {
        "raise_exception": MovedPermanentlyError,
        "message": "The resource you are looking for is moved to another URL."
    },
    304: {
        "raise_exception": NotModifiedError,
        "message": "The requested resource has not been modified since the last time you accessed it."
    },
    400:{
        "raise_exception": BadRequestException,
        "message": "The request is missing or has a bad parameter."
    },
    401: {
        "raise_exception": BadCredentialsException,
        "message": "Invalid authorization credentials."
    },
    403: {
        "raise_exception": AuthException,
        "message": "User doesn't have permission to access the resource."
    },
    404: {
        "raise_exception": NotFoundException,
        "message": "The resource you have specified cannot be found. Alternatively the access_token is not valid for the resource"
    },
    409: {
        "raise_exception": ConflictError,
        "message": "The request could not be completed due to a conflict with the current state of the server."
    },
    422: {
        "raise_exception": UnprocessableError,
        "message": "The request was not able to process right now."
    },
    429: {
        "raise_exception": TooManyRequests,
        "message": "Too many requests occurred."
    },
    500: {
        "raise_exception": InternalServerError,
        "message": "An error has occurred at Github's end."
    }
}

def raise_for_error(resp, source, stream, client, should_skip_404):
    """
    Retrieve the error code and the error message from the response and return custom exceptions accordingly.
    """
    error_code = resp.status_code
    try:
        response_json = resp.json()
    except JSONDecodeError:
        response_json = {}

    if error_code == 404 and should_skip_404:
        # Add not accessible stream into list.
        client.not_accessible_repos.add(stream)
        details = ERROR_CODE_EXCEPTION_MAPPING.get(error_code).get("message")
        if source == "teams":
            details += ' or it is a personal account repository'
        message = "HTTP-error-code: 404, Error: {}. Please refer \'{}\' for more details.".format(details, response_json.get("documentation_url"))
        LOGGER.warning(message)
        # Don't raise a NotFoundException
        return None

    message = "HTTP-error-code: {}, Error: {}".format(
        error_code, ERROR_CODE_EXCEPTION_MAPPING.get(error_code, {}).get("message", "Unknown Error") if response_json == {} else response_json)

    if error_code > 500:
        raise Server5xxError(message) from None

    exc = ERROR_CODE_EXCEPTION_MAPPING.get(error_code, {}).get("raise_exception", GithubException)
    raise exc(message) from None

def calculate_seconds(epoch):
    """
    Calculate the seconds to sleep before making a new request.
    """
    current = time.time()
    return int(round((epoch - current), 0))

def rate_throttling(response, max_sleep_seconds):
    """
    For rate limit errors, get the remaining time before retrying and calculate the time to sleep before making a new request.
    """
    if 'X-RateLimit-Remaining' in response.headers:
        if int(response.headers['X-RateLimit-Remaining']) == 0:
            seconds_to_sleep = calculate_seconds(int(response.headers['X-RateLimit-Reset']))

            if seconds_to_sleep > max_sleep_seconds:
                message = "API rate limit exceeded, please try after {} seconds.".format(seconds_to_sleep)
                raise RateLimitExceeded(message) from None

            LOGGER.info("API rate limit exceeded. Tap will retry the data collection after %s seconds.", seconds_to_sleep)
            time.sleep(seconds_to_sleep)
    else:
        # Raise an exception if `X-RateLimit-Remaining` is not found in the header.
        # API does include this key header if provided base URL is not a valid github custom domain.
        raise GithubException("The API call using the specified base url was unsuccessful. Please double-check the provided base URL.")

class GithubClient:
    """
    The client class used for making REST calls to the Github API.
    """
    def __init__(self, config):
        self.config = config
        self.session = requests.Session()
        self.base_url = config['base_url'] if config.get('base_url') else DEFAULT_DOMAIN
        self.max_sleep_seconds = self.config.get('max_sleep_seconds', DEFAULT_SLEEP_SECONDS)
        self.set_auth_in_session()
        self.not_accessible_repos = set()

    def get_request_timeout(self):
        """
        Get the request timeout from the config, if not present use the default 300 seconds.
        """
        # Get the value of request timeout from config
        config_request_timeout = self.config.get('request_timeout')

        # Only return the timeout value if it is passed in the config and the value is not 0, "0" or ""
        if config_request_timeout and float(config_request_timeout):
            return float(config_request_timeout)

        # Return default timeout
        return REQUEST_TIMEOUT

    def set_auth_in_session(self):
        """
        Set access token in the header for authorization.
        """
        access_token = self.config['access_token']
        self.session.headers.update({'authorization': 'token ' + access_token})

    # pylint: disable=dangerous-default-value
    # During 'Timeout' error there is also possibility of 'ConnectionError',
    # hence added backoff for 'ConnectionError' too.
    @backoff.on_exception(backoff.expo, (requests.Timeout, requests.ConnectionError, Server5xxError, TooManyRequests), max_tries=5, factor=2)
    def authed_get(self, source, url, headers={}, stream="", should_skip_404 = True):
        """
        Call rest API and return the response in case of status code 200.
        """
        with metrics.http_request_timer(source) as timer:
            self.session.headers.update(headers)
            resp = self.session.request(method='get', url=url, timeout=self.get_request_timeout())
            if resp.status_code != 200:
                raise_for_error(resp, source, stream, self, should_skip_404)
            timer.tags[metrics.Tag.http_status_code] = resp.status_code
            rate_throttling(resp, self.max_sleep_seconds)
            if resp.status_code == 404:
                # Return an empty response body since we're not raising a NotFoundException
                resp._content = b'{}' # pylint: disable=protected-access
            return resp

    def authed_get_all_pages(self, source, url, headers={}, stream="", should_skip_404 = True):
        """
        Fetch all pages of records and return them.
        """
        while True:
            r = self.authed_get(source, url, headers, stream, should_skip_404)
            yield r

            # Fetch the next page if next found in the response.
            if 'next' in r.links:
                url = r.links['next']['url']
            else:
            # Break the loop if all pages are fetched.
                break

    def verify_repo_access(self, url_for_repo, repo):
        """
        Call rest API to verify that the user has sufficient permissions to access this repository.
        """
        try:
            self.authed_get("verifying repository access", url_for_repo)
        except NotFoundException:
            # Throwing user-friendly error message as it checks token access
            message = "HTTP-error-code: 404, Error: Please check the repository name \'{}\' or you do not have sufficient permissions to access this repository.".format(repo)
            raise NotFoundException(message) from None

    def verify_access_for_repo(self):
        """
        For all the repositories mentioned in the config, check the access for each repos.
        """
        repositories, org = self.extract_repos_from_config() # pylint: disable=unused-variable

        for repo in repositories:

            url_for_repo = "{}/repos/{}/commits".format(self.base_url, repo)
            LOGGER.info("Verifying access of repository: %s", repo)

            # Verifying for Repo access
            self.verify_repo_access(url_for_repo, repo)

    def extract_orgs_from_config(self):
        """
        Extracts all organizations from the config
        """
        repo_paths = list(filter(None, self.config['repository'].split(' ')))
        orgs_paths = [repo.split('/')[0] for repo in repo_paths]

        return set(orgs_paths)

    def extract_repos_from_config(self):
        """
        Extracts all repositories from the config and calls get_all_repos()
        for organizations using the wildcard 'org/*' format.
        """
        repo_paths = list(filter(None, self.config['repository'].split(' ')))

        unique_repos = set()
        # Insert the duplicate repos found in the config repo_paths into duplicates
        duplicate_repos = [x for x in repo_paths if x in unique_repos or (unique_repos.add(x) or False)]
        if duplicate_repos:
            LOGGER.warning("Duplicate repositories found: %s and will be synced only once.", duplicate_repos)

        repo_paths = list(set(repo_paths))

        orgs_with_all_repos = []
        orgs = []
        repos_with_errors = []
        for repo in repo_paths:
            # Split the repo_path by `/` as we are passing org/repo_name in the config.
            split_repo_path = repo.split('/')
            # Prepare list of organizations
            orgs.append(split_repo_path[0])
            # Check for the second element in the split list only if the length is greater than 1 and the first/second
            # element is not empty (for scenarios such as: `org/` or `/repo` which is invalid)
            if len(split_repo_path) > 1 and split_repo_path[1] != '' and split_repo_path[0] != '':
                # If the second element is *, we need to check access for all the repos.
                if split_repo_path[1] == '*':
                    orgs_with_all_repos.append(repo)
            else:
                # If `/`/repo name/organization not found, append the repo_path in the repos_with_errors
                repos_with_errors.append(repo)

        # If any repos found in repos_with_errors, raise an exception
        if repos_with_errors:
            raise GithubException("Please provide valid organization/repository for: {}".format(sorted(repos_with_errors)))

        if orgs_with_all_repos:
            # Remove any wildcard "org/*" occurrences from `repo_paths`
            repo_paths = list(set(repo_paths).difference(set(orgs_with_all_repos)))

            # Get all repositories for an org in the config
            all_repos = self.get_all_repos(orgs_with_all_repos)

            # Update repo_paths
            repo_paths.extend(all_repos)

        return repo_paths, set(orgs)

    def get_all_repos(self, organizations: list):
        """
        Retrieves all repositories for the provided organizations and
        verifies basic access for them.

        Docs: https://docs.github.com/en/rest/reference/repos#list-organization-repositories
        """
        repos = []

        for org_path in organizations:
            org = org_path.split('/')[0]
            try:
                for response in self.authed_get_all_pages(
                    'get_all_repos',
                    '{}/orgs/{}/repos?sort=created&direction=desc'.format(self.base_url, org),
                    should_skip_404 = False
                ):
                    org_repos = response.json()
                    LOGGER.info("Collected repos for organization: %s", org)

                    for repo in org_repos:
                        repo_full_name = repo.get('full_name')
                        LOGGER.info("Verifying access of repository: %s", repo_full_name)

                        self.verify_repo_access(
                            '{}/repos/{}/commits'.format(self.base_url,repo_full_name),
                            repo
                        )

                        repos.append(repo_full_name)
            except NotFoundException:
                # Throwing user-friendly error message as it checks token access
                message = "HTTP-error-code: 404, Error: Please check the organization name \'{}\' or you do not have sufficient permissions to access this organization.".format(org)
                raise NotFoundException(message) from None

        return repos

    def __exit__(self, exception_type, exception_value, traceback):
        # Kill the session instance.
        self.session.close()
